File: services/api/src/domain/listItem.ts
import { z } from "zod";

export const ListTypeSchema = z.enum(["todo", "watch", "later", "antiTodo"]);
export type ListType = z.infer<typeof ListTypeSchema>;

export const ListStatusSchema = z.enum([
  "planned",
  "in_progress",
  "completed",
  "archived"
]);
export type ListStatus = z.infer<typeof ListStatusSchema>;

export const ListItemSchema = z.object({
  id: z.string().uuid(),
  title: z.string().min(1),
  listType: ListTypeSchema,
  status: ListStatusSchema.default("planned"),
  notes: z.string().max(2000).optional(),
  dueAt: z.string().datetime().optional(),
  followUpAt: z.string().datetime().optional(),
  createdAt: z.string().datetime(),
  completedAt: z.string().datetime().nullable().optional(),
  tags: z.array(z.string()).default([]),
  confidenceScore: z.number().min(0).max(1).default(0.5)
});
export type ListItem = z.infer<typeof ListItemSchema>;

export const NewListItemSchema = ListItemSchema.partial({
  id: true,
  createdAt: true,
  status: true,
  tags: true,
  confidenceScore: true
})
  .extend({
    title: z.string().min(1),
    listType: ListTypeSchema
  })
  .transform((payload) => ({
    id: payload.id ?? crypto.randomUUID(),
    createdAt: payload.createdAt ?? new Date().toISOString(),
    status: payload.status ?? "planned",
    tags: payload.tags ?? [],
    confidenceScore: payload.confidenceScore ?? 0.5,
    ...payload
  }));
export type NewListItem = z.infer<typeof NewListItemSchema>;

export const UpdateListItemSchema = ListItemSchema.partial({
  id: true,
  createdAt: true
});
export type UpdateListItem = z.infer<typeof UpdateListItemSchema>;

File: apps/ios/AndreApp/Sources/AndreApp/Models/ListItem.swift
import Foundation

public struct ListItem: Identifiable, Hashable, Codable {
    public enum ListType: String, Codable, CaseIterable {
        case todo
        case watch
        case later
        case antiTodo

        public var displayName: String {
            switch self {
            case .todo: return "Todo"
            case .watch: return "Watch"
            case .later: return "Later"
            case .antiTodo: return "Anti-Todo"
            }
        }
    }

    public enum Status: String, Codable {
        case planned
        case inProgress
        case completed
        case archived
    }

    public let id: UUID
    public var title: String
    public var listType: ListType
    public var status: Status
    public var notes: String?
    public var dueAt: Date?
    public var followUpAt: Date?
    public var createdAt: Date
    public var completedAt: Date?
    public var tags: [String]

    public init(
        id: UUID = UUID(),
        title: String,
        listType: ListType,
        status: Status = .planned,
        notes: String? = nil,
        dueAt: Date? = nil,
        followUpAt: Date? = nil,
        createdAt: Date = .now,
        completedAt: Date? = nil,
        tags: [String] = []
    ) {
        self.id = id
        self.title = title
        self.listType = listType
        self.status = status
        self.notes = notes
        self.dueAt = dueAt
        self.followUpAt = followUpAt
        self.createdAt = createdAt
        self.completedAt = completedAt
        self.tags = tags
    }
}

extension ListItem {
    static let placeholderItems: [ListItem] = [
        .init(title: "Ship API design draft", listType: .todo, dueAt: .now.addingTimeInterval(86_400)),
        .init(title: "Follow up with Sarah on partnership", listType: .watch, followUpAt: .now.addingTimeInterval(172_800)),
        .init(title: "Research calendar integration", listType: .later, notes: "Check GCAL API quotas")
    ]
}
